src/plan/barriers.rs

```diff
@@ -78,7 +78,15 @@ impl<E: ProcessEdgesWork> ObjectRememberingBarrier<E> {
                 Ordering::SeqCst,
                 Ordering::SeqCst,
             ) {
+                // We just logged the object
                 return true;
+            } else {
+                let old_value =
+                    load_metadata::<E::VM>(&self.meta, object, None, Some(Ordering::SeqCst));
```

Comment:
Doing a test when CAS fails here does not look like a normal way people would write CAS loops. For example this is how jikesrvm does the compare and swap: https://github.com/JikesRVM/JikesRVM/blob/5072f19761/MMTk/src/org/mmtk/policy/LargeObjectSpace.java#L242-L247

I think this is due to the API design problem of the side metadata. `compare_exchange_metadata` (L72 above) operates on bits not bytes, and the inner implementation already performs an extra load. But after the load, it does not check if the loaded value is already in the correct state:
https://github.com/mmtk/mmtk-core/blob/42262c296e164b5b1d05103e527ce5a9aee7cfd2/src/util/metadata/side_metadata/global.rs#L471-L484

This is the fundamental problem to force us to do an extra load here. With this problem, either the old approach (do the load and test before CAS) or the new approach (do the load after CAS) will always cause extra loads. And also notice that the code for metadata address computation will be repeated as well, because for this case `self.meta` is not a compile-time constant, and the metadata code can hardly be inlined and optimized.

The ideal solution would be either (1) removing the load in  `compare_exchange_metadata`, or (2) doing the check inside `compare_exchange_metadata`. Alternatively, (3) we can add a rust-std-style [`fetch_update`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_update) interface to simplify the barrier CAS loop:

```rust
side_metadata::fetch_update(&self.meta, Ordering::SeqCst, , Ordering::SeqCst, |old_bit: usize| {
    if old_bit == 0 {
        None
    } else {
        Some(1)
    }
});
```

This `fetch_update` function will also ensure that the metadata address will only be compiled once, regardless of how the compiler does its optimization.

I actually have the 2nd and the 3rd option implemented in my lxr branch:
[1] https://github.com/wenyuzhao/mmtk-core/blob/lxr-merge/src/util/metadata/side_metadata/global.rs#L590-L592
[2] https://github.com/wenyuzhao/mmtk-core/blob/lxr-merge/src/util/metadata/side_metadata/global.rs#L734-L803

Reply:
Thanks for the comments. Both 2 and 3 look good. I would suggest that I will do that change after this pull request. Also I found that for `bits_num_log == 6`, our side metadata uses `AtomicUsize` rather than `AtomicU64` -- that should be a bug. I will do a PR to fix both issues.

Can you let me know for this PR, should I revert this change, or just keep it? My intention is to turn the unconditional load into a conditional load (only load after CAS fails).

Reply:
I prefer 3.  Because the metadata may be finer than a byte, any attempt to provide a plain `compare_exchange_metadata` API inevitably have to involve an extra load inside the function, so as not to modify surrounding bits, unless the function also takes the old byte as the parameter, which is ugly.  Designing a `fetch_update` API will also allow us to use the underlying `AtomicU8.fetch_update` in the Rust standard library.  In theory, this allows the compiler to implement it with load-link store-conditional (LL-SC) instead of compare-exchange on some architectures (almost anything except x86). I don't know if the Rust compiler already does this, but it is possible.

Another alternative is to provide `fetch_and_metadata`, `fetch_or_metadata`, `fetch_xor_metadata`.  Some architectures provide special instructions for atomic bit operations that never fails, and therefore never need to be used in a loop.  In Rust, there is [`AtomicU8.fetch_and`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU8.html#method.fetch_and), and such methods are [available in the `atomic_traits` crate](https://docs.rs/atomic-traits/latest/atomic_traits/fetch/trait.And.html), too.

Reply:
Yes it makes sense to do metadata API refactoring in another PR.

> Can you let me know for this PR, should I revert this change, or just keep it?

Please keep it. This looks more correct. To me, it doesn't make sense for the old implementation to load the bit once for a quick test and load it again for CAS.

Reply:
I actually attempted 3. The `metadata` field in the object remembering barrier is actually `Metadata` rather than `SideMetadata` (because the unlogged bit could be in the object header). So it means if we add `fetch_update`, we need to add it to the `ObjectModel` trait and allow bindings to implement it for in-header metadata. For that reason, I thought it was better to put it to a separate pull request.

I will submit a pull request for that.

Reply:
I tried compiling `fetch_and` and `fetch_update` to AArch64 and RISCV64.  `fetch_and` always results in simpler code without loop.  `fetch_update` is implemented as a CAS loop on AArch64, but implemented as an LR/SC loop on RISCV64.

So the message is very clear.  If we only do bit-set and bit-clear, use `fetch_or` and `fetch_and`;  Otherwise, prefer `fetch_update` over a hand-written CAS loop.